mmengine/utils/progressbar_rich.py

```diff
@@ -0,0 +1,180 @@
+
+import rich
+from rich.progress import Progress
+
+from ..utils import is_seq_of
+from .timer import Timer
+
+
+class RichProgressBar:
+    """use rich to enhance progress bar."""
+
+    def __init__(self):
+        self.bar = Progress()
+        self.tasks = []
+        self.descriptions = []
+        self.colors = []
+        self.timer = None
+        self.completed = 0
+        self.infinite = False
+
+        self.bar.start()
+
+    def __del__(self):
+        self.bar.stop()
+
+    @staticmethod
+    def write(msg: str, color: str = 'blue'):
+        rich.print(f'[{color}]{msg}')
+
+    def add_task(self,
+                 total: int = None,
+                 color: str = 'blue',
+                 description: str = 'Process...') -> int:
+        if total is not None:
+            assert not self.infinite, (
+                'The prior task is an infinite task (total is None), '
+                'RichProgressBar can only accept one infinite task')
+
+            if total <= 0:
+                raise ValueError(
+                    'Total only exists if it is greater than zero or None.')
+            self.tasks.append(
+                self.bar.add_task(
+                    f'[{color}]{description}_0/{total}', total=total))
+            self.colors.append(color)
+            self.descriptions.append(description)
+            task_id = len(self.tasks) - 1
+        else:
+            assert not self.tasks, (
+                'Since the total argument is None, this task is considered '
+                'infinite. RichProgressBar should not have any other tasks '
+                'added to it when an infinite task is present.')
+            self.write('completed: 0, elapsed: 0s', color)
+            self.timer = Timer()
+            self.infinite = True
+            self.completed = 0
+            self.tasks.append(0)
+            self.colors.append(color)
+            self.descriptions.append(description)
+            task_id = 0
+        return task_id
+
+    def update(self, task_id: int = 0, advance: int = 1):
+        if advance <= 0:
+            raise ValueError('advance should greater than zero.')
+
+        if self.infinite:
+            self.completed += advance
+            elapsed = self.timer.since_start()
+            if elapsed > 0:
+                fps = self.completed / elapsed
+            else:
+                fps = float('inf')
+            self.write(
+                f'completed: {self.completed}, '
+                f'elapsed: {int(elapsed + 0.5)}s, '
+                f'{fps:.1f} tasks/s', self.colors[task_id])
+        else:
+            assert task_id >= -1 and task_id < len(self.tasks), (
+                'The task_id must be within a valid range')
+
+            if task_id == -1:
+                task_id = len(self.tasks) - 1
+
+            completed = self.bar.tasks[task_id].completed + advance
+            total = self.bar.tasks[task_id].total
+            self.bar.update(
+                task_id,
```

Comment:
Instead of using a simple task index, should `task_id` be determined based on the result of `self.bar.add_task`?
BTW, should `add_task` also return the result of `self.bar.add_task`?